### PR TITLE
Breaking: rework to make it parity with gotrue-js

### PR DESCRIPTION
**Pull request by dshukertjr**
## What kind of change does this PR introduce?

Make this library parity with [gotrue-js v2.0 rc](https://github.com/supabase/gotrue-js/tree/rc). 

The biggest difference is that with this PR, we have two namespaces:`client` and `client.admin`. 

`client` is used for any properties/ methods used within the client side where as any methods in `client.admin` is meant to be used on the server side. 

## What is the current behavior?

Some of the methods don't match what is in gotrue-js v2

## What is the new behavior?

The list of methods match. There are few missing methods on the `GotrueAdmin` class, but those are meant to be server side methods, so we can add them later on.

## Additional context
So sorry for this long PR. I should have split this one into smaller ones. There are a lot of moving parts in this PR, especially in the public API, so any suggestions/ comments are welcome!

Also, I haven't been able to bring over all the test cases from gotrue-js, so if anyone wants to add more tests to this one, feel free to add onto this PR!

Open id connect related methods have been removed both from this library and the js library, but they will be added back as soon as we sort out the server side issues. 

Also, js library uses `client.user()` and `client.session()` to get user and session, but I have this library with `client.currentUser` and `client.currentSession`. The reason for this is:
- Firebase used `currentUser` and `currentSession`
- js SDK's `user()` and `session()` methods are async
- We have removed `.user()` and `.session()` v1.0.0-dev.1 of gotrue-dart

,but let me know if we should stick with this, or should we go back to `.user()` and `.session()`.

### CODE REVIEW

**Comment by dshukertjr** [comment]
Also, now that supabase-js v2.0 is officially released, I'd like to release a stable v1.0 of supabase-flutter once this change is merged, but is there anything else that we have left to do while it is still in developer preview?

File: lib/src/types/auth_response.dart
Comment (bdlukaa) [inline lib/src/types/auth_response.dart:25]:
```suggestion
  const OAuthResponse({
```

File: lib/src/types/auth_response.dart
Comment (bdlukaa) [inline lib/src/types/auth_response.dart:48]:
```suggestion
  const AuthSessionUrlResponse({
```

File: lib/src/types/auth_response.dart
Comment (bdlukaa) [inline lib/src/types/auth_response.dart:97]: this should be private  _or_  not exported

**Review by bdlukaa — CHANGES_REQUESTED** [review]

File: pubspec.yaml
Comment (Vinzent03) [inline pubspec.yaml:8]: Do we really need to increase this? The static method on `GenerateLinkType` can be created with `extensions` as well.

File: lib/src/gotrue_admin_api.dart
Comment (Vinzent03) [inline lib/src/gotrue_admin_api.dart:23]: Can't we simply pass the headers to the `GotrueRequestOptions`?

**Comment by Vinzent03** [comment]
Suggested edit:
```diff
diff --git a/lib/src/types/auth_exception.dart b/lib/src/types/auth_exception.dart
index 8d7f98b..318b162 100644
--- a/lib/src/types/auth_exception.dart
+++ b/lib/src/types/auth_exception.dart
@@ -6,7 +6,7 @@ class AuthException implements Exception {
 
   @override
   String toString() =>
-      'GoTrueException(message: $message, statusCode: $statusCode)';
+      'AuthException(message: $message, statusCode: $statusCode)';
 
   @override
   bool operator ==(Object other) {

```

**Review by Vinzent03 — COMMENTED** [review]
I would change the `toString()` of `AuthException`. https://github.com/supabase-community/gotrue-dart/pull/90/files#diff-76d7358a19549a6171b6bb6988a09b2c72efcc4a5e74020ad148b8bfa11e6866R9

File: lib/src/types/auth_response.dart
Comment (dshukertjr) [reply lib/src/types/auth_response.dart:97]: Great call!

File: pubspec.yaml
Comment (bdlukaa) [reply pubspec.yaml:8]: This will enable other features we may want to use in the future that are only available in the latest dart versions. Also, this is a major version release, so the breaking change makes sense.

Comment (dshukertjr) [reply pubspec.yaml:8]: Good call. I will bring it back to 2.12. I think other libraries use 2.17 as well, so I will look into them and bring them down to 2.12.

File: lib/src/gotrue_admin_api.dart
Comment (dshukertjr) [reply lib/src/gotrue_admin_api.dart:24]: Maybe I'm missing something, but don't I created the variable `headers` so that we can override the `Authorization` header in line 23, so I don't think we can pass the headers directly to `GotrueRequestOptions` here.

File: lib/src/gotrue_client.dart
Comment (dshukertjr) [inline lib/src/gotrue_client.dart:186]: In js client, this method is called `signInWithOAuth`, but I have renamed it `getOAuthSignInUrl`, because all this method is doing is getting the URL. I want to rename the [`signInWithProvider()`](https://github.com/supabase-community/supabase-flutter/blob/main/lib/src/supabase_auth.dart#L303) to `signInWithOAuth()`, which I think is more aligned with the js SDK, but what does everyone think about that?

File: lib/src/gotrue_admin_api.dart
Comment (Vinzent03) [reply lib/src/gotrue_admin_api.dart:24]: Oh I've written the wrong. I meant we can pass the `jwt` to `GoTrueRequestOptions`, because it overrides the header [here](https://github.com/supabase-community/gotrue-dart/pull/90/files#diff-fb71a9f29effc766e3d7775f310b2b560d7abe17048f0280e5ce605517d983f1R46) by itself.

File: pubspec.yaml
Comment (Vinzent03) [reply pubspec.yaml]: I think the question is, how long do we want to support old Dart versions. I don't know if everyone is already using Flutter >=3.0.0, which doesn't ship Dart 1.17, right?

Comment (dshukertjr) [reply pubspec.yaml]: Sorry @bdlukaa, some how I missed your message. The enhanced enum in 2.17.0 is enthusing and I personally use it all the time, but 2.17.0 was released only 5 months ago, and I feel like bumping the SDK version up to it, even though I first tried to do it in this PR, would cut off pretty good chunk of users. I think 2.14.0 is a good one to settle on, because it was released more than a year ago, and it at least has enhanced typedef feature.

File: lib/src/gotrue_admin_api.dart
Comment (dshukertjr) [reply lib/src/gotrue_admin_api.dart:24]: @Vinzent03 Ah, amazing eye! I totally missed that. Pushing a fix now!

**Comment by Vinzent03** [comment]
Shouldnt `runtimeType` be included in the `hashcode` from `AuthException`? Two different classes with the same properties would have the same `hashcode`.

**Comment by dshukertjr** [comment]
@Vinzent03 
I think that part is fine. Quote from [Flutter docs](https://api.flutter.dev/flutter/dart-core/Object/hashCode.html):

>Objects that are not equal are allowed to have the same hash code. It is even technically allowed that all instances have the same hash code, but if clashes happen too often, it may reduce the efficiency of hash-based data structures like [HashSet](https://api.flutter.dev/flutter/dart-collection/HashSet-class.html) or [HashMap](https://api.flutter.dev/flutter/dart-collection/HashMap-class.html).

**Comment by dshukertjr** [comment]
@supabase-community/dart-maintainers 
I was wondering if there are more things that should be changed/ fixed on this one, but are there any?

**Review by Vinzent03 — COMMENTED** [review]
I think the code is fine, but from how I understood tests, there are very anti pattern, because you depend on test A to be run before test B. I'm pretty sure the should be able to be run individually.

In general, why do we use callbacks for `onAuthStateChange`? Why not use `Streams`, which are more Dart style and into the language integrated. I think most people will convert them to a stream nevertheless.

File: test/admin_test.dart
Comment (dshukertjr) [reply test/admin_test.dart:54]: Also, I haven't looked into it too much, but it seems like the tests of the JS SDK runs independently, so we could probably take it from them. For now though, I prefer if we merge this one, release v1.0 and then work on optimizing the tests!

**Comment by dshukertjr** [comment]
@Vinzent03 
Changing `onAuthStateChange` into `Stream` sounds great. Let's see what I can whip up.

**Comment by dshukertjr** [comment]
With this update, onAuthStateChange is now a `Stream` and can be used like this. Not sure if `AuthState` is the best name for it, so am open for suggestions if there are any!

```dart
final StreamSubscription<AuthState> subscription = supabase.auth.onAuthStateChange.listen((event) {
  final AuthChangeEvent event = event.event;
  final Session? session = event.session;
  // Do something with event and session
});
```

File: test/admin_test.dart
Comment (Vinzent03) [reply test/admin_test.dart:42]: Ok, just for ref: https://github.com/supabase-community/supabase-flutter/issues/236

File: lib/src/gotrue_client.dart
Comment (Vinzent03) [inline lib/src/gotrue_client.dart:237]: `singInWithOtp` never returns a `user` or `session`, can't this method be of type `Future<void>` then?

**Comment by Vinzent03** [comment]
I think the name is fine. I thought about `AuthEvent`, but that's already somewhat used for the event type (`AuthEvent.event`).
Despite that (hopefully) last comment: LGTM!

**Review by Vinzent03 — APPROVED** [review]

File: lib/src/gotrue_client.dart
Comment (dshukertjr) [reply lib/src/gotrue_client.dart:237]: Makes sense! Let me add that update and merge this PR 🎉

**Pull request merged**
